LibZstd Clang.jl generated bindings

mkitti opened the pull request:
This pull requests creates a new submodule LibZstd that is generated by Clang.jl. It uses the new Clang.jl 0.14 interface from the master branch. Thanks to @Gnimuc for overhauling

The generator is run by checking out the master branch of Clang.jl, and then executing `libzstd/generator.jl`. It has options specified in `libzstd/generator.toml`. This includes a manual patch this is applied by generator.jl. The existing libzstd.jl is modified to use the new LibZstd submodule.

This is a step towards addressing #28 to expose the dictionary training API of ZSTD via Julia bindings.

yuyichao commented on src/libzstd.jl:
```diff
@@ -89,19 +87,21 @@ mutable struct DStream
 
 function initialize!(dstream::DStream)
-    return ccall((:ZSTD_initDStream, libzstd), Csize_t, (Ptr{Cvoid},), dstream.ptr)
+    return LibZstd.ZSTD_initDStream(dstream.ptr)
```
Do NOT do these object to pointer conversions manually. Leave it to ccall to do it. I’m assuming clang isn’t generating a signature that requires you to do such conversion manually. If that’s not the case then the code it generate is broken and must be fixed. ( a brief look suggests that the generated wrappers are correct)

same with all uses of pointer_from_objref here.

yuyichao replied:
Actually I just realized that the alignment made me think that the change is new to this pr but it’s actually not….


I that case it should still be fixed but doesn’t have to be done here.

mkitti replied:
I'm unclear what you are suggesting here.

`LibZstd.ZSTD_initDStream` now calls `ccall`. All I'm doing with the change above is consolidating all the `ccall` references into the `LibZstd` module.
https://github.com/JuliaIO/CodecZstd.jl/blob/c084472af8cf40d30cab56001f866d4700ac0e31/src/libzstd/LibZstd.jl#L307-L309

Perhaps you mean that I should just pass `dstream` directly rather than `dstream.ptr` or something like that?

yuyichao replied:
Yes the change is fine. But both the new and the old version should have the Julia object (dsteam in this case) directly ccall. You’ll need to define the correct unsafe convert to make it work.

mkitti replied:
I see what you mean. This probably has some garbage collection implications as well now that I think about it.

I think I might want to save those changes for another PR and perhaps another release.

yuyichao replied:
The only reason it’s broken is indeed because of object lifetime.

bicycle1885 commented:
Thank you for your contribution!

Is it possible to include Project.toml and Manifest.toml for the generator in src/libzstd so that we can quickly regenerate LibZstd.jl?

mkitti commented:
Yes, I think that is possible with one caveat. Currently,  Clang.jl version 0.14 is not released or registered. However, there is an active campaign to convert the existing generator scripts. Perhaps @Gnimuc can advise here.

mkitti commented:
> Thank you for your contribution!
> 
> Is it possible to include Project.toml and Manifest.toml for the generator in src/libzstd so that we can quickly regenerate LibZstd.jl?

In ca42fe4, I added Project.toml, Manifest.toml, as well some comments and documentation about how to run the generator.

mkitti commented:
My recommendation is to release this as `v0.7.1` since none of these changes should be breaking.

Gnimuc commented:
> Yes, I think that is possible with one caveat. Currently, Clang.jl version 0.14 is not released or registered. However, there is an active campaign to convert the existing generator scripts. Perhaps @Gnimuc can advise here.

It's always safe to use `Project.toml`&`Manifest.toml` to pin a specific version of an environment, as long as one does not accidentally update the `Manifest.toml`. Clang.jl is not a dependency of this package, so it doesn't matter whether a released version is being used or not.

mkitti commented:
Hi @bicycle1885 , I added the files as requested. I've also addressed other reviewer's concerns in next PR. Is there anything we should do before merging?

mkitti commented:
Are there any objections to merging this? I would like to merge by Monday October 4th if there are no objections.

mkitti commented:
If there are no other objections, I will merge this in several hours. Thank you for the reviews.
